pkg/controlplane/controller/defaultservice/default_service_controller.go

```diff
@@ -0,0 +1,350 @@
+		}
+		c.endpointReconciler.Destroy()
+	}()
+
+	select {
+	case <-finishedReconciling:
+		// done
+	case <-time.After(2 * c.endpointInterval):
+		// don't block server shutdown forever if we can't reach etcd to remove ourselves
+		klog.Warning("RemoveEndpoints() timed out")
+	}
+}
+
+// PostStartHook initiates the core controller loops that must exist for bootstrapping.
+func (c *Controller) PostStartHook(hookContext genericapiserver.PostStartHookContext) error {
+	go c.start(hookContext.StopCh)
+	return nil
+}
+
+// PreShutdownHook triggers the actions needed to shut down the API Server cleanly.
+func (c *Controller) PreShutdownHook() error {
+	c.stopOnce.Do(c.stop)
+	return nil
+}
+
+// sync attempts to update the default Kube service and the endpoint records
+func (c *Controller) sync() error {
+	if err := c.createNamespaceIfNeeded(); err != nil {
+		return err
+	}
+
+	if err := c.createOrUpdateMasterServiceIfNeeded(); err != nil {
+		return err
+	}
+	endpointPorts := createEndpointPortSpec(c.publicServicePort, "https")
+	if err := c.endpointReconciler.ReconcileEndpoints(serviceName, c.publicIP, endpointPorts, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createPortAndServiceSpec creates an array of service ports.
+// If the NodePort value is 0, just the servicePort is used, otherwise, a node port is exposed.
+func createPortAndServiceSpec(servicePort int, targetServicePort int, nodePort int, servicePortName string) ([]v1.ServicePort, v1.ServiceType) {
+	// Use the Cluster IP type for the service port if NodePort isn't provided.
+	// Otherwise, we will be binding the master service to a NodePort.
+	servicePorts := []v1.ServicePort{{
+		Protocol:   v1.ProtocolTCP,
+		Port:       int32(servicePort),
+		Name:       servicePortName,
+		TargetPort: intstr.FromInt(targetServicePort),
+	}}
+	serviceType := v1.ServiceTypeClusterIP
+	if nodePort > 0 {
+		servicePorts[0].NodePort = int32(nodePort)
+		serviceType = v1.ServiceTypeNodePort
+	}
+	return servicePorts, serviceType
+}
+
+// createEndpointPortSpec creates the endpoint ports
+func createEndpointPortSpec(endpointPort int, endpointPortName string) []v1.EndpointPort {
+	return []v1.EndpointPort{{
+		Protocol: v1.ProtocolTCP,
+		Port:     int32(endpointPort),
+		Name:     endpointPortName,
+	}}
+}
+
+func (c *Controller) createNamespaceIfNeeded() error {
+	if _, err := c.client.CoreV1().Namespaces().Get(context.TODO(), metav1.NamespaceDefault, metav1.GetOptions{}); err == nil {
+		// the namespace already exists
+		return nil
+	}
+	newNs := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      metav1.NamespaceDefault,
+			Namespace: "",
+		},
+	}
+	_, err := c.client.CoreV1().Namespaces().Create(context.TODO(), newNs, metav1.CreateOptions{})
+	if err != nil && errors.IsAlreadyExists(err) {
+		err = nil
+	}
+	return err
+}
+
+// createOrUpdateMasterServiceIfNeeded will create the specified service if it doesn't already exist.
```

Comment:
```suggestion
// createOrUpdateMasterServiceIfNeeded will create the specified service if it doesn't already exist.
// TODO: rename to avoid use of term "master"
```